sql/2020/11_Security/feature_adoption_by_other_features.sql

```diff
@@ -0,0 +1,33 @@
+#standardSQL
+# Analysis of how certain features influence the adoption of other features
+SELECT
+  _TABLE_SUFFIX AS client,
+  headername,
+  COUNT(0) AS total_pages,
+  COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername))) AS total_with_header,
+  COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername))) / COUNT(0) AS pct,
+  COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)) AND STARTS_WITH(url, 'https')) / COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername))) AS pct_header_and_https,
+  COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)) AND REGEXP_CONTAINS(respOtherHeaders, '(?i)Content-Security-Policy')) / COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername))) AS pct_header_and_csp,
```

Comment:
Same issue here and I actually ran this to see what the difference would be and got this (1st line is CSP, 2nd is CSP-RO):

![CSP usage](https://user-images.githubusercontent.com/10931297/98804089-9e10ad80-240d-11eb-8f76-14b4241b478a.png)

So not an insignificant difference. It's only 93.8% usage when you're saying it's 100%.

Also interesting that 10% of CSP sites use CSP-RO.
And everyone using CSP-RO is also using CSP - is no one testing any more and everyone has gone all in?

Also think we should do a SAFE_DIVIDE here. Doesn't matter for now, but some of these are very small and could go to zero in near future (I experienced this when trying to run this with a different UNNEST that had a typo!)

Here's what I ended up doing:

```
#standardSQL
# Analysis of how certain features influence the adoption of other features
SELECT
  _TABLE_SUFFIX AS client,
  headername,
  COUNT(0) AS total_pages,
  COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername))) AS total_with_header,
  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername))), COUNT(0)) AS pct,
  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)) AND STARTS_WITH(url, 'https')), COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)))) AS pct_header_and_https,
  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)) AND REGEXP_CONTAINS(respOtherHeaders, '(?i)Content-Security-Policy')), COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)))) AS pct_header_and_csp,
  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)) AND REGEXP_CONTAINS(respOtherHeaders, '(?i)Content-Security-Policy ')), COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)))) AS pct_header_and_csp_space,
  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)) AND REGEXP_CONTAINS(respOtherHeaders, '(?i)Content-Security-Policy[^-]')), COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)))) AS pct_header_and_csp_not_dash,
  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)) AND REGEXP_CONTAINS(respOtherHeaders, '(?i)Content-Security-Policy-Report-Only')), COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)))) AS pct_header_and_csp_report_only,
  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)) AND REGEXP_CONTAINS(respOtherHeaders, '(?i)Cross-Origin-Embedder-Policy')), COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)))) AS pct_header_and_coep,
  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)) AND REGEXP_CONTAINS(respOtherHeaders, '(?i)Cross-Origin-Opener-Policy')), COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)))) AS pct_header_and_coop,
  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)) AND REGEXP_CONTAINS(respOtherHeaders, '(?i)Cross-Origin-Resource-Policy')), COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)))) AS pct_header_and_corp,
  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)) AND REGEXP_CONTAINS(respOtherHeaders, '(?i)Expect-CT')), COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)))) AS pct_header_and_expectct,
  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)) AND REGEXP_CONTAINS(respOtherHeaders, '(?i)Feature-Policy')), COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)))) AS pct_header_and_featurep,
  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)) AND REGEXP_CONTAINS(respOtherHeaders, '(?i)Permissions-Policy')), COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)))) AS pct_header_and_permissionsp,
  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)) AND REGEXP_CONTAINS(respOtherHeaders, '(?i)Referrer-Policy')), COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)))) AS pct_header_and_referrerp,
  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)) AND REGEXP_CONTAINS(respOtherHeaders, '(?i)Report-To')), COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)))) AS pct_header_and_reportto,
  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)) AND REGEXP_CONTAINS(respOtherHeaders, '(?i)Strict-Transport-Security')), COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)))) AS pct_header_and_hsts,
  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)) AND REGEXP_CONTAINS(respOtherHeaders, '(?i)X-Content-Type-Options')), COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)))) AS pct_header_and_xcto,
  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)) AND REGEXP_CONTAINS(respOtherHeaders, '(?i)X-Frame-Options')), COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)))) AS pct_header_and_xfo,
  SAFE_DIVIDE(COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)) AND REGEXP_CONTAINS(respOtherHeaders, '(?i)X-XSS-Protection')), COUNTIF(REGEXP_CONTAINS(respOtherHeaders, CONCAT('(?i)', headername)))) AS pct_header_and_xss,
FROM
  `httparchive.summary_requests.2020_08_01_*` AS r,
UNNEST(['Content-Security-Policy', 'Content-Security-Policy-Report-Only', 'Cross-Origin-Embedder-Policy', 'Cross-Origin-Opener-Policy', 'Cross-Origin-Resource-Policy', 'Expect-CT', 'Feature-Policy', 'Permissions-Policy', 'Referrer-Policy', 'Report-To', 'Strict-Transport-Security', 'X-Content-Type-Options', 'X-Frame-Options', 'X-XSS-Protection']) AS headername
WHERE
  firstHtml
GROUP BY
  client,
  headername
ORDER BY
  client,
  headername
```

Reply:
Hmm, perhaps we should do the same for all the headers (that is, to put a space behind it in the regex)

Here the difference between CSP and CSP-RO doesn't matter so much I think; the 1.0 just shows that for hosts that use CSP or CSP-RO, all of them have CSP or CSP-RO - it does matter for the other queries though

Is there a particular reason to include `Content-Security-Policy[^-]`?

Reply:
> Is there a particular reason to include `Content-Security-Policy[^-]`?

No. Just me testing if space or dash was better separator. Let's use space since by above that's better as can use in other headers.